homotopy-core/src/graph.rs

```diff
@@ -1,310 +1,295 @@
     Boundary(Boundary),
     /// From a diagram's cospans.
     Internal(SingularHeight, Direction),
     /// Sparse identity from a rewrite.
-    Sparse,
+    Sparse(RegularHeight),
+    /// Unit slices from a rewrite.
+    UnitSlice,
     /// Cone regular slices from a rewrite.
     RegularSlice,
-    /// Cone `singular_slices` (slice `x` of `y`) from a rewrite.
-    SingularSlice(SingularHeight, usize),
+    /// Cone singular slices from a rewrite.
+    SingularSlice(SingularHeight),
 }
 
-pub fn explode<OldNodeKey, NewNodeKey, OldEdgeKey, NewEdgeKey, NodeMap, EdgeMap, S, T>(
-    graph: &SliceGraph<OldNodeKey, OldEdgeKey, S>,
-    node_map: NodeMap,
-    edge_map: EdgeMap,
-) -> Result<
-    (
-        SliceGraph<NewNodeKey, NewEdgeKey, T>,
-        IdxVec<NodeIndex<S>, Vec<Option<NodeIndex<T>>>>,
-    ),
-    DimensionError,
->
+/// The output of explosion consists of the exploded graph and some maps from the original graph into the exploded graph.
+#[derive(Clone, Debug)]
+pub struct ExplosionOutput<V, E, Ix1, Ix2>
 where
-    OldNodeKey: Clone + std::fmt::Debug,
-    OldEdgeKey: Clone + std::fmt::Debug,
-    NodeMap: Fn(NodeIndex<S>, SliceIndex, OldNodeKey) -> Option<NewNodeKey>,
-    EdgeMap: Fn(Option<EdgeIndex<S>>, RewriteOrigin, Option<OldEdgeKey>) -> Option<NewEdgeKey>,
-    S: IndexType,
-    T: IndexType,
+    Ix1: IndexType,
+    Ix2: IndexType,
 {
-    use Height::{Regular, Singular};
-
-    let mut exploded_graph: SliceGraph<NewNodeKey, NewEdgeKey, T> = Default::default();
-
-    // Maps every node in the original graph to its slices in the exploded graph.
-    let mut node_to_slices: IdxVec<NodeIndex<S>, Vec<Option<NodeIndex<T>>>> =
-        IdxVec::with_capacity(graph.node_count());
-
-    for (n, (key, diagram)) in graph.node_references() {
-        let diagram: &DiagramN = diagram.try_into()?;
-
-        let mut slices = Vec::with_capacity(diagram.size() * 2 + 3);
-
-        // Source slice
-        slices.push((
-            node_map(n, Boundary::Source.into(), key.clone()),
-            diagram.source(),
-        ));
-
-        // Interior slices
-        for (i, slice) in diagram.slices().enumerate() {
-            slices.push((node_map(n, Height::from(i).into(), key.clone()), slice));
-        }
-
-        // Target slice
-        slices.push((
-            node_map(n, Boundary::Target.into(), key.clone()),
-            diagram.target(),
-        ));
-
-        let nodes: Vec<Option<NodeIndex<T>>> = slices
-            .into_iter()
-            .map(|(k, d)| k.map(|i| exploded_graph.add_node((i, d))))
-            .collect();
-
-        // Identity rewrite from source slice
-        if let (Some(s), Some(t)) = (nodes[0], nodes[1]) {
-            if let Some(key) = edge_map(None, RewriteOrigin::Boundary(Boundary::Source), None) {
-                exploded_graph.add_edge(s, t, (key, Rewrite::identity(diagram.dimension() - 1)));
-            }
-        }
+    pub output: SliceGraph<V, E, Ix2>,
```

Comment:
I explicitly chose not to use `graph` because that way it could be confused with the original graph. I think `output` makes it clearer this is the *output* of explosion. What do you mean about the type being opaque?